doc/logfilter.1.rst
===========
 logfilter
===========

--------------------------------------------
filter some logs based on date and log level
--------------------------------------------

.. include:: include/metadata.rst
   :start-line: 1

:Version: logfilter |version|
:Date: |date|
:Manual section: 1

SYNOPSIS
========

**logfilter** [**-h**] [**-a** *DATE*] [**-b** *DATE*] [**--batch** \|
**--no-batch**] [**-l** *LEVEL*] [*FILE* ...]

DESCRIPTION
===========

**logfilter** prints lines from log files which match the given date
range and severity level.
The goal is to produce a digestible screenful of the most recent and
most urgent log messages.

OPTIONS
=======

-h, --help            Print a short help message and exit.

-a DATE, --after=DATE
    Filter logs older than *DATE*.
    *DATE* can be any date string understood by the ``--date`` option
    of GNU Coreutils or Coreutils-compatible **date**\ (1).

-b DATE, --before=DATE
    Filter logs newer than *DATE*.
    The format of *DATE* is the same as for ``--after``, above.

--batch, --no-batch   Don’t print headers giving file names.

-l LEVEL, --level=LEVEL
    Filter logs below *LEVEL*.
    *LEVEL* can be one of
    {EMERG, ALERT, CRITICAL, ERROR, WARNING, NOTICE, INFO, DEBUG}
    or its abbreviation.

CONFIGURATION
=============

Create a config file to customize the default values of the script
variables. These values can be overridden by command-line arguments. The
config file should be a series of variable–value pairs, separated by an
equals sign (‘=’), each on their own line. Whitespace is ignored, and
lines starting with ‘#’ will be ignored as comments.

The XDG configuration directory search path will be searched for config
files named ‘logfilter/config’. The search path is based on the values
of environment variables XDG_CONFIG_DIRS and XDG_CONFIG_HOME.
XDG_CONFIG_HOME will be searched first (if unset then this is equal to
‘~/.config’), then XDG_CONFIG_DIRS (if unset then this is equal to
‘/etc/xdg’). Values in earlier files take precedence over values in
later files. Most regular users should just put their config file at
‘~/.config/logfilter/config’.

Variables:

after : date
    Default value of ``--after``
before : date
    Default value of ``--before``
batch : Boolean
    Default setting of ``--batch``
datefmt : format string
    Format of datestamps in log files
level : string
    Default setting of ``--level``
logfiles : path names
    Default value of *FILE* (used if no *FILE*\ s are specified at the
    command line)
program : string
    Program text used by AWK
progfile : path name
    Name of file containing AWK program to use instead of *program*

Built-in defaults:

.. include:: include/defaults.txt
   :literal:

See **date**\ (1) for the accepted format of date and format string
arguments.

Per-logfile configuration
-------------------------

In addition to files named ‘logfilter/config’, the configuration
directory search path will be searched for files named
‘logfilter/logfiles.conf’. This is the per-logfile configuration file.
It contains variable–value pairs like the config file described above,
but each setting must be preceded by a section name in square brackets.
It might look something like this:

::

   [*app.log]
   datefmt = +%Y-%m-%dT%H:%M:%S
   program = $1 > after && $1 <= before && $2 ~ level

If a logfile argument *FILE* matches a section name in a per-logfile
configuration file, then the variable settings in that section will be
used for that file. The remainder of the settings will be taken from
‘config’ or from the special default section DEFAULT. In the above
example, any logfile matching the filename pattern ‘\*app.log’ will use
the values of **datefmt** and **program** from that section.
If multiple section names match a *FILE* name, settings from later
matching sections will override settings from earlier sections.
So, as an example, a section named ‘\*’ will override settings in all
preceding sections, because ‘\*’ will match any name.

See **glob**\ (7) for a description of wildcard matching pathnames.

Setting the variables **logfiles** and **batch** in a section other than
DEFAULT has no effect, because these are program-wide settings.

FILES
=====

*/etc/xdg/logfilter/*\ {*config*,\ *logfiles.conf*}
    System configuration files

*~/.config/logfilter/*\ {*config*,\ *logfiles.conf*}
    User configuration files


logfilter.py
#!/usr/bin/python3
#
#  Copyright 2023–2025 Dylan Maltby
#  SPDX-Licence-Identifier: Apache-2.0
#
"""Filter some logs using AWK.

Depends on:
-  AWK
-  Coreutils date
"""

from __future__ import annotations

import argparse
import collections
import configparser
import fnmatch
import glob
import logging
import os
import shlex
import shutil
import subprocess
import sys
import types
from collections.abc import (
    Callable,
    Iterable,
    Iterator,
    Mapping,
    MutableMapping,
    Sequence,
)
from typing import Any, NoReturn, Optional, Union

__prog__ = "logfilter"
__version__ = "0.4.0"

CONFIG_PATH = "config"
LOGFILES_CONF_PATH = "logfiles.conf"
# LOGGING_DISPLAY_NAMES
LOG_LEVELS: list[str] = [
    "EMERG",
    "ALERT",
    "CRITICAL",
    "ERROR",
    "WARNING",
    "NOTICE",
    "INFO",
    "DEBUG",
]
DEFAULTS: dict[str, str] = {
    "after": "today-3days",
    "before": "today+1day",
    "batch": "0",
    "datefmt": "+%Y-%m-%d",
    "level": "INFO",
    "logfiles": "~/.log/*.log",
    "program": "$1 > after && $1 <= before && $3 ~ level",
}
BOOLEAN_STATES: dict[str, bool] = {
    "1": True,
    "yes": True,
    "true": True,
    "on": True,
    "0": False,
    "no": False,
    "false": False,
    "off": False,
}

# Argument type for cmds of subprocess.run
Arg = Union[str, bytes, os.PathLike[str], os.PathLike[bytes]]


# pylint: disable=too-few-public-methods
class Args(types.SimpleNamespace):
    """Namespace for parsed command-line arguments"""

    logfiles: list[str]
    after: str | None
    before: str | None
    batch: bool
    level: str | None


def disambiguate(
    names: Iterable[str], func: Callable[[str], str] = str
) -> Callable[[str], str]:
    """Return a function that can disambiguate a string between *names*.

    >>> disambiguate(LOG_LEVELS, str.upper)("warn")
    'WARNING'
    """

    def type_checker(value: str) -> str:
        value = func(value)
        candidates = [name for name in names if func(name).startswith(value)]
        if len(candidates) != 1:
            return value
        return candidates[0]

    return type_checker


def expand_paths(paths: str) -> Iterator[str]:
    """
    Split *paths* like a shell, and expand environment variables and globs.
    """
    for word in shlex.split(paths):
        word = os.path.expanduser(word)
        word = os.path.expandvars(word)
        yield from glob.iglob(word)


def convert_boolean(value: str) -> bool:
    return BOOLEAN_STATES.get(value.lower(), bool(value))


def build_cla_parser(defaults: Mapping[str, str]) -> argparse.ArgumentParser:
    """Build and return command-line argument parser."""
    parser = argparse.ArgumentParser(
        prog=__prog__, description="Filter some logs based on date and log level."
    )
    logfiles = defaults["logfiles"]
    parser.add_argument(
        "logfiles",
        nargs="*",
        default=list(expand_paths(logfiles)),
        metavar="FILE",
        help=f"filter %(metavar)s(s) or default logfiles: {logfiles}",
    )
    parser.add_argument(
        "-a",
        "--after",
        metavar="DATE",
        help="filter logs older than %(metavar)s",
    )
    parser.add_argument(
        "-b",
        "--before",
        metavar="DATE",
        help="filter logs newer than %(metavar)s",
    )
    parser.set_defaults(batch=convert_boolean(defaults["batch"]))
    parser.add_argument(
        "--batch",
        action=argparse.BooleanOptionalAction,
        help="don't print headers giving file names",
    )
    parser.add_argument(
        "-l",
        "--level",
        type=disambiguate(LOG_LEVELS, str.upper),
        choices=LOG_LEVELS,
        metavar="LEVEL",
        help="filter logs below %(metavar)s (choose from %(choices)s)",
    )
    return parser


def main(argv: Sequence[str] | None = None) -> None:
    """Parsing arguments from *argv*, print results to sys.stdout."""
    if os.environ.get("LF_DEBUG"):
        logging.basicConfig(level=logging.DEBUG)
    general_defaults = load_defaults(DEFAULTS)
    cfg = configparser.ConfigParser(defaults=general_defaults, interpolation=None)
    try:
        read_configuration(cfg, LOGFILES_CONF_PATH)
    except configparser.Error as err:
        die(f"Error with configuration file: {err}")
    cfg_defaults = cfg.defaults()
    args = build_cla_parser(cfg_defaults).parse_args(argv, namespace=Args())
    logging.debug(args)
    logfiles = args.logfiles
    if not logfiles:
        logging.debug("null glob")
        return

    for logfile in logfiles:
        section = get_matching_settings(logfile, cfg) or cfg_defaults
        try:
            awk_variables = _set_awk_variables(args, section)
            awk_options = _set_awk_options(section)
        except configparser.Error as err:
            die(f"Error with configuration file: {err}")
        if not args.batch:
            print()
            print("==>", logfile, "<==")
            # Flush headers before awk starts writing
            sys.stdout.flush()
        awk(files=[logfile], **awk_options, variables=awk_variables)


def _set_awk_variables(args: Args, section: Mapping[str, str]) -> dict[str, str]:
    awk_variables = {}
    level = args.level or section["level"]
    try:
        awk_variables["level"] = "|".join(LOG_LEVELS[: LOG_LEVELS.index(level) + 1])
    except ValueError:
        section_name = getattr(section, "name", "DEFAULT")
        die(f"In section [{section_name}]: invalid level name: {level}")
    awk_variables["after"] = datestr(args.after or section["after"], section["datefmt"])
    awk_variables["before"] = datestr(
        args.before or section["before"], section["datefmt"]
    )
    return awk_variables


def _set_awk_options(section: Mapping[str, str]) -> dict[str, str]:
    if progfile := section.get("progfile"):
        return {"progfiles": progfile}
    return {"program_text": section["program"]}


# pylint: disable=too-many-arguments
def awk(
    files: Iterable[Arg],
    program_text: Optional[Arg] = None,
    *,
    progfiles: Optional[Iterable[Arg]] = None,
    variables: Optional[Mapping[Any, Any]] = None,
    field_sep: Optional[Arg] = None,
    executable: Union[str, os.PathLike[str]] = "awk",
) -> int:
    """Call `awk` with the given arguments, returning its exit status.

    ::
        awk [-v variables...] [-F field_sep] [-f progfiles... | program_text]
            [files...]
    """
    exe_path = shutil.which(os.environ.get("LF_AWK", executable)) or die(
        f"{executable}: command not found"
    )
    cmds: list[Arg] = [exe_path]
    if variables is not None:
        for var, value in variables.items():
            cmds += ["-v", f"{var}={value}"]
    if field_sep is not None:
        cmds += ["-F", field_sep]
    if program_text is not None:
        cmds += ["--", program_text, *files]
    elif progfiles is not None:
        for program_file in progfiles:
            cmds += ["-f", program_file]
        cmds += ["--", *files]
    logging.debug(cmds)
    try:
        proc = subprocess.run(cmds, check=True)
    except subprocess.CalledProcessError as err:
        die(str(err))
    return proc.returncode


def load_config_paths(*resource: Union[str, os.PathLike[str]]) -> Iterator[str]:
    """
    Return an iterator which gives each directory named *resource* in the
    configuration search path. Information provided by earlier directories
    should take precedence over later ones.

    Procedure copied from:
    https://pyxdg.readthedocs.io/en/latest/_modules/xdg/BaseDirectory.html
    """
    xdg_config_home = os.environ.get("XDG_CONFIG_HOME") or os.path.join(
        os.path.expanduser("~"), ".config"
    )
    xdg_config_dirs = [
        xdg_config_home,
        *(os.environ.get("XDG_CONFIG_DIRS") or "/etc/xdg").split(":"),
    ]
    resource_path = os.path.join(*resource)
    for config_dir in xdg_config_dirs:
        path = os.path.join(config_dir, resource_path)
        if os.path.exists(path):
            yield path


def load_defaults(defaults: MutableMapping[str, str]) -> collections.ChainMap[str, str]:
    """Merge *defaults* with k:v loaded from configuration files."""
    maps = []
    for cfg in load_config_paths(__prog__, CONFIG_PATH):
        try:
            file = open(cfg, encoding="utf-8")
        except OSError:
            logging.debug("found config file but couldn't open for reading: %s", cfg)
            continue
        except UnicodeDecodeError as err:
            logging.debug(
                "found config file but couldn't decode as UTF-8: %s: %s", err, cfg
            )
            continue
        with file:
            maps.append(parse_kv_config(file))
            logging.debug("read configuration from file: %s", cfg)
    return collections.ChainMap(*maps, defaults)


def read_configuration(
    config: configparser.ConfigParser, pathname: Union[str, os.PathLike[str]]
) -> list[str]:
    """Load configuration files into *config*.

    Return a list of files read.
    """
    cfgs = list(load_config_paths(__prog__, pathname))
    files_read = config.read(reversed(cfgs))
    logging.debug("read configuration from files: %s", files_read)
    return files_read


def get_matching_settings(
    name: str, config: configparser.ConfigParser
) -> collections.ChainMap[str, str]:
    """Return settings from sections in *config* that fnmatch *name*.

    If there are multiple sections that match *name*, settings from later
    sections override those from earlier sections.
    """
    # Reverse sections so that they are in the correct order for ChainMap,
    # i.e. the _later_ matching sections are placed _first_ the ChainMap
    # lookup sequence.
    sections = (
        config[section]
        for section in reversed(config.sections())
        if fnmatch.fnmatch(name, section)
    )
    return collections.ChainMap(*sections)


def parse_kv_config(reader: Iterable[str]) -> dict[str, str]:
    """Return a dict of keys to values parsed from lines of text in *reader*.

    Simple syntax:
      - Keys separated from values by '='
      - Comments starting with '#'
      - Each line either a key–value pair or a comment
      - External whitespace ignored
    """
    symbols = {}
    for line in reader:
        line = line.lstrip()
        if line.startswith("#"):
            continue
        key, sep, val = line.partition("=")
        if not sep:
            continue
        symbols[key.strip().lower()] = val.strip()
    return symbols


def datestr(date: Optional[str] = None, datefmt: Optional[str] = None) -> str:
    """Call `date` with the given arguments and return its stdout as a string.

    ::
        $(date [--date ${date}] [${datefmt}])
    """
    executable = shutil.which("date") or die("date: command not found")
    cmds = [executable]
    if date is not None:
        cmds += ["--date", date]
    if datefmt is not None:
        cmds.append(datefmt)
    try:
        proc = subprocess.run(cmds, check=True, stdout=subprocess.PIPE)
    except subprocess.CalledProcessError as err:
        die(str(err))
    return proc.stdout.decode().strip()


def die(message: str) -> NoReturn:
    print(f"{__prog__}: {message}", file=sys.stderr)
    sys.exit(1)


if __name__ == "__main__":
    main()
